File: basedatabindingcomponents/src/main/java/com/tanveershafeeprottoy/basedatabindingcomponents/adapters/BaseSpinnerAdapter.java
package com.tanveershafeeprottoy.basedatabindingcomponents.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import com.tanveershafeeprottoy.basedatabindingcomponents.viewholders.BaseSpinnerAdapterViewHolder;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;

/**
 * @author Tanveer Shafee Prottoy
 */
public class BaseSpinnerAdapter<T> extends ArrayAdapter<T> {
    private int resourceId;
    private int variableId;
    private List<T> objList;

    public BaseSpinnerAdapter(Context context, int resourceId, int variableId, List<T> objList) {
        super(context, resourceId, objList);
        this.resourceId = resourceId;
        this.variableId = variableId;
        this.objList = objList;
    }

    @Override
    @NonNull
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        return getRowView(position, convertView, parent);
    }

    @Override
    public View getDropDownView(int position, View convertView, @NonNull ViewGroup parent) {
        return getRowView(position, convertView, parent);
    }

    @Override
    public int getCount() {
        try {
            return objList.size();
        }
        catch(NullPointerException n) {
            return 0;
        }
    }

    private View getRowView(int position, View convertView, ViewGroup parent) {
        BaseSpinnerAdapterViewHolder baseSpinnerAdapterViewHolder;
        if(convertView == null) {
            ViewDataBinding viewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), resourceId, parent, false);
            baseSpinnerAdapterViewHolder = new BaseSpinnerAdapterViewHolder(viewDataBinding);
            convertView = viewDataBinding.getRoot();
            convertView.setTag(baseSpinnerAdapterViewHolder);
        }
        else {
            baseSpinnerAdapterViewHolder = (BaseSpinnerAdapterViewHolder) convertView.getTag();
        }
        baseSpinnerAdapterViewHolder.getViewDataBinding().setVariable(variableId, objList.get(position));
        return convertView;
    }
}

File: basedatabindingcomponents/src/main/java/com/tanveershafeeprottoy/basedatabindingcomponents/adapters/BaseSpinnerAdapter.kt
package com.tanveershafeeprottoy.basedatabindingcomponents.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.tanveershafeeprottoy.basedatabindingcomponents.viewholders.BaseSpinnerAdapterViewHolder

class BaseSpinnerAdapter(context: Context, private val resourceId: Int, textViewId: Int, private val variableId: Int, private val anyList: List<Any>) : ArrayAdapter<Any>(context, resourceId, textViewId, anyList) {

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return getRowView(position, convertView, parent)
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return getRowView(position, convertView, parent)
    }

    override fun getCount(): Int {
        return try {
            anyList.size
        }
        catch(n: NullPointerException) {
            0
        }
    }

    private fun getRowView(position: Int, convertView: View?, parent: ViewGroup): View {
        var rowView = convertView
        val baseSpinnerAdapterViewHolder: BaseSpinnerAdapterViewHolder
        if(rowView == null) {
            val viewDataBinding: ViewDataBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.context), resourceId, parent, false)
            baseSpinnerAdapterViewHolder = BaseSpinnerAdapterViewHolder(viewDataBinding)
            rowView = viewDataBinding.root
            rowView.tag = baseSpinnerAdapterViewHolder
        }
        else {
            baseSpinnerAdapterViewHolder = rowView.tag as BaseSpinnerAdapterViewHolder
        }
        baseSpinnerAdapterViewHolder.viewDataBinding.setVariable(variableId, anyList[position])
        return rowView
    }
}